### PR TITLE
Added video unpublish support

### LINKED ISSUE #2362: Video Search: make offered_by optional for videos
We would like to not show an offered_by filter in the ui for some videos. We will update the code to accept playlist with null offered_by in the  youtube.yaml file in the open-video-data repo

Epic:
https://github.com/mitodl/open-discussions/issues/2302

### LINKED ISSUE #2363: Video Search: mark videos no longer in channel playlists as published=False
This supports a few use cases:

- A video is unpublished/deleted from youtube itself
- A video is removed from the playlist(s)
- A playlist if removed from the list of playlists we sync, effectively removing those videos unless they show up in another list that is synced

This should update videos for that channel that didn't show up in the sync that just occurred and mark them `is_published=False`. We probably need to take some steps to ensure that if there's an error (error retrieving youtube data, error saving to DB, error indexing the video) that we don't erroneously mark videos as unpublished.

### PR DESCRIPTION
**Pull request by rhysyngsun**
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes #2362
Closes #2363
Part of #2367 

#### What's this PR do?
- Makes `offered_by` optional in the configuration
- Adds tracking of channels and playlists (this will enable us to easily generate user lists from this data for #2367 in a subsequent PR)
- Adds support for unpublishing videos (the logic here is that if a video no longer appears in any playlist we're tracking we consider it unpublished and should not appear in search)

#### How should this be manually tested?

- Delete all video records
- Run `./manage.py backpopulate_youtube_data fetch`
- Verify you have 5988 OCW videos and 54 mitx videos on the search page
- In `.env`, set `OPEN_VIDEO_DATA_BRANCH=nl/small_playlist`, restart workers
- Run `./manage.py backpopulate_youtube_data fetch`
- Verify you have 34 OCW videos and 54 mitx videos on the search page

### CODE REVIEW

**Comment by rhysyngsun** [comment]
Found some issues in the data, looking into that

**Comment by mbertrand** [comment]
Can't test functionality right now because of this error, but I'll look over the code.  Could this be a problem on production too or can we increase the daily quota?

```
HttpError 403 when requesting https://www.googleapis.com/youtube/v3/channels.... "Daily Limit Exceeded. The quota will be reset at midnight Pacific Time (PT). You may monitor your quota usage and adjust limits in the API Console: 
```

File: course_catalog/etl/loaders.py [hunk 1]
```diff
@@ -212,5 +224,160 @@ def load_video(video_data):
...
+    # video_channels_data is a generator
+    for video_channel_data in video_channels_data:
+        try:
+            video_channel = load_video_channel(video_channel_data)
+        except ExtractException:
+            # video_channel_data has lazily evaluated generators, one of them could raise an extraction error
+            # this is a small pollution of separation of concerns
+            # but this allows us to stream the extracted data w/ generators
+            # as opposed to having to load everything into memory, which will eventually fail
+            log.exception("Error with extracted video channel")
```
Comment (mbertrand) [inline course_catalog/etl/loaders.py]: Would it be useful to display part of the `video_channel_data` in the exception message to narrow down which one caused a problem?

File: course_catalog/etl/youtube.py [hunk 2]
```diff
@@ -34,56 +41,139 @@ def get_youtube_client():
...
+        raise ExtractPlaylistItemException(
+            "Error fetching playlist items: playlist_id={playlist_id}"
+        ) from exc
 
-            if "nextPageToken" in playlist_items_response:
-                page_token = playlist_items_response["nextPageToken"]
-            else:
+
+def extract_playlists(youtube_client, playlist_configs):
+    """Extract a list of playlists"""
```
Comment (mbertrand) [inline course_catalog/etl/youtube.py]: Add `Args` and `Returns`

File: course_catalog/etl/youtube.py [hunk 3]
```diff
@@ -34,56 +41,139 @@ def get_youtube_client():
...
+    except StopIteration:
+        return
+    except googleapiclient.errors.HttpError as exc:
+        raise ExtractPlaylistException(
+            "Error fetching channel playlists: playlist_ids={playlist_ids}"
+        ) from exc
+
+
+def extract_channels(youtube_client, channels_config):
+    """Extract a list of channels"""
```
Comment (mbertrand) [inline course_catalog/etl/youtube.py]: Add `Args` and `Returns`

File: course_catalog/etl/loaders.py [hunk 4]
```diff
@@ -212,5 +224,160 @@ def load_video(video_data):
...
+                )
+            )
+        )
+        .filter(published=True)
+        .exclude(in_published_playlist=True)
+    ):
+        # remove it from the index first
+        search_task_helpers.delete_video(video)
+        video.published = False
+        video.save()
```
Comment (mbertrand) [inline course_catalog/etl/loaders.py]: Looks like these 3 lines aren't covered in unit tests

File: course_catalog/etl/youtube.py [hunk 5]
```diff
@@ -150,6 +240,74 @@ def github_youtube_config_files():
...
+            list of errors or an empty list if no errors
+    """
+    errors = []
+
+    if not channel_config:
+        errors.append("Channel config data is empty")
+        return errors
+
+    if not isinstance(channel_config, dict):
+        errors.append("Channel data should be a dict")
```
Comment (mbertrand) [inline course_catalog/etl/youtube.py]: cover this in a unit test?

File: course_catalog/etl/youtube.py [hunk 6]
```diff
@@ -150,6 +240,74 @@ def github_youtube_config_files():
...
+        return errors
+
+    for required_key in ["playlists", "channel_id"]:
+        if required_key not in channel_config:
+            errors.append(f"Required key '{required_key}' is not present")
+
+    for idx, playlist_config in enumerate(channel_config.get("playlists", [])):
+        if "id" not in playlist_config:
+            errors.append(f"Required key 'id' not present in playlists[{idx}]")
```
Comment (mbertrand) [inline course_catalog/etl/youtube.py]: add a unit test?

**Comment by mbertrand** [comment]
Functionality works as advertised, with 113 bootcamp videos too.

**Comment by rhysyngsun** [comment]
@mbertrand feedback addressed

**Pull request merged**